src/rust/lib/graph-editor/src/component/visualization/data.rs

```diff
@@ -0,0 +1,107 @@
+//! This module defines the `Data` struct and related functionality.
+
+use crate::prelude::*;
+
+use crate::component::visualization::EnsoType;
+
+use serde::Deserialize;
+
+
+
+// ======================================
+// === Wrapper for Visualisation Data ===
+// =======================================
+
+/// Type indicator
+pub type DataType = EnsoType;
+
+/// Wrapper for data that can be consumed by a visualisation.
+/// TODO[mm] consider static versus dynamic typing for visualizations and data!
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub enum Data {
+    JSON   { content : Rc<serde_json::Value> },
+    Binary { content : Rc<dyn Any>           },
+}
+
+impl Data {
+    /// Returns the data as as JSON. If the data cannot be returned as JSON, it will return a
+    /// `DataError` instead.
+    pub fn as_json(&self) -> Result<Rc<serde_json::Value>, DataError> {
+        match &self {
+            Data::JSON { content } => Ok(Rc::clone(content)),
+            _ => { Err(DataError::InvalidDataType{})  },
+        }
+    }
+
+    /// Returns the wrapped data in Rust format. If the data cannot be returned as rust datatype, a
+    /// `DataError` is returned instead.
+    pub fn as_binary<T>(&self) -> Result<Rc<T>, DataError>
+        where for<'de> T:Deserialize<'de> + 'static {
+        match &self {
+            Data::JSON { content } => {
+                // We try to deserialize here. Just in case it works.
+                // This is useful for simple data types where we don't want to care to much about
+                // representation, e.g., a list of numbers.
+                let value : serde_json::Value = content.as_ref().clone();
+                if let Ok(result) = serde_json::from_value(value) {
+                    Ok(Rc::new(result))
+                } else {
+                    Err(DataError::InvalidDataType)
+                }
+            },
+            Data::Binary { content } => { Rc::clone(content).downcast()
+                .or(Err(DataError::InvalidDataType))},
+        }
+    }
```

Comment:
Oh, now I realized what you are doing here. So here is one problem. The binary data stream does not mean that it is a binary data understandable by Rust. It is a binary data encoded by Enso function (can be user function which defines the binary layout!) and it should be just received as bytes by visualization. In this format, parsing and understanding the binary data is up to the visualization author. Almost never this data would be castable to rust types.